File: dat-pypi/tests/test_bench.py
import asyncio
import secrets
import string
import time
import unittest
from concurrent.futures import ThreadPoolExecutor

from saro_dat import DatSignatureAlgorithm, DatCryptoAlgorithm, DatCertificate, DatManager, DatSignature, DatCrypto


def generate_base62(length: int) -> str:
    characters = string.ascii_letters + string.digits
    return ''.join(secrets.choice(characters) for _ in range(length))


def run_issue(cert, plain, secure):
    return DatManager._issue(cert, plain, secure)


def run_parse(cert, dat_str):
    return DatManager._parse(cert, dat_str)


async def loops(multi_thread: bool, loop_size: int, certificates: list[DatCertificate], plain: str, secure: str):
    mode_name = "Multi-Thread" if multi_thread else "Single-Thread"
    print(f"\n--- {mode_name} ---")

    for cert in certificates:
        pre = f"{cert._signature_key.algorithm.value} {cert._crypto_key.algorithm.value}"

        start = time.perf_counter()
        last_dat = ""

        if multi_thread:
            # 파이썬의 GIL 우회를 위해 ProcessPoolExecutor 사용 (실제 멀티코어 활용)
            with ThreadPoolExecutor() as executor:
                futures = [executor.submit(run_issue, cert, plain, secure) for _ in range(loop_size)]
                for fut in futures:
                    last_dat = fut.result()
        else:
            for _ in range(loop_size):
                last_dat = DatManager._issue(cert, plain, secure)

        duration_ms = (time.perf_counter() - start) * 1000
        print(f"{pre} Issue * {loop_size} : {duration_ms:.0f}ms")

        # 2. Parse Benchmark
        start = time.perf_counter()
        last_payload = None

        if multi_thread:
            with ThreadPoolExecutor() as executor:
                futures = [executor.submit(run_parse, cert, last_dat) for _ in range(loop_size)]
                for fut in futures:
                    last_payload = fut.result()
        else:
            for _ in range(loop_size):
                last_payload = DatManager._parse(cert, last_dat)

        duration_ms = (time.perf_counter() - start) * 1000
        print(f"{pre} Parse * {loop_size} : {duration_ms:.0f}ms")

        # 검증
        assert last_payload.plain == plain
        assert last_payload.secure == secure


async def benchmark(loop_size: int):
    plain = generate_base62(100)
    secure = generate_base62(100)

    print("Performance Test (Plain, Secure)")
    print(f"Plain: {plain}")
    print(f"Secure: {secure}")

    certificates = []
    now = int(time.time())

    for sa in DatSignatureAlgorithm:
        for ca in DatCryptoAlgorithm:
            certificates.append(DatCertificate(0, DatSignature.generate(sa), DatCrypto.generate(ca), now - 10, now + 600, 60))

    await loops(True, loop_size, certificates, plain, secure)
    await loops(False, loop_size, certificates, plain, secure)


class TestBench(unittest.TestCase):
    def test(self):
        loop_size = 10000
        asyncio.run(benchmark(loop_size))


if __name__ == "__main__":
    unittest.main()

File: dat-ruby/test/test_bench.rb
# frozen_string_literal: true

require_relative './test_helper'
require 'benchmark'
require 'parallel'
require 'etc'

class TestBench < Minitest::Test
  def generate_base62(length)
    characters = ('a'..'z').to_a + ('A'..'Z').to_a + ('0'..'9').to_a
    Array.new(length) { characters.sample }.join
  end

  def loops(multi_thread, loop_size, certificates, plain, secure)
    mode_name = multi_thread ? "Multi-Thread" : "Single-Thread"
    puts "\n--- #{mode_name} ---"

    processors_count = Etc.nprocessors

    certificates.each do |cert|
      pre = "#{cert.signature_key.algorithm} #{cert.crypto_key.algorithm}"

      start_time = Time.now
      last_dat = ""

      if multi_thread
        results = Parallel.map(1..loop_size, in_processes: processors_count) do |_i|
          Saro::Dat::DatManager._issue(cert, plain, secure)
        end
        last_dat = results.last
      else
        loop_size.times do
          last_dat = Saro::Dat::DatManager._issue(cert, plain, secure)
        end
      end

      duration_ms = (Time.now - start_time) * 1000
      puts "#{pre} Issue * #{loop_size} : #{duration_ms.to_i}ms"


      # 2. Parse Benchmark
      start_time = Time.now
      last_payload = nil

      if multi_thread
        results = Parallel.map(1..loop_size, in_processes: processors_count) do |_i|
          Saro::Dat::DatManager._parse(cert, last_dat)
        end
        last_payload = results.last
      else
        loop_size.times do
          last_payload = Saro::Dat::DatManager._parse(cert, last_dat)
        end
      end

      duration_ms = (Time.now - start_time) * 1000
      puts "#{pre} Parse * #{loop_size} : #{duration_ms.to_i}ms"

      # 검증
      assert_equal plain, last_payload.plain
      assert_equal secure, last_payload.secure
    end
  end

  def test_bench
    loop_size = 10000
    plain = generate_base62(100)
    secure = generate_base62(100)

    puts "Performance Test (Plain, Secure)"
    puts "Plain: #{plain}"
    puts "Secure: #{secure}"

    certificates = []
    now = Time.now.to_i

    [
      Saro::Dat::DatSignatureAlgorithm::HMAC_SHA256_MFS,
      Saro::Dat::DatSignatureAlgorithm::HMAC_SHA384_MFS,
      Saro::Dat::DatSignatureAlgorithm::HMAC_SHA512_MFS,
      Saro::Dat::DatSignatureAlgorithm::ECDSA_P256,
      Saro::Dat::DatSignatureAlgorithm::ECDSA_P384,
      Saro::Dat::DatSignatureAlgorithm::ECDSA_P521
    ].each do |sa|
      Saro::Dat::DatCryptoAlgorithm.all.each do |ca|
        begin
          certificates << Saro::Dat::DatCertificate.new(
            0,
            now - 10,
            610,
            60,
            Saro::Dat::DatSignature.generate(sa),
            Saro::Dat::DatCrypto.generate(ca)
          )
        rescue OpenSSL::Cipher::CipherError
        end
      end
    end

    loops(true, loop_size, certificates, plain, secure)
    loops(false, loop_size, certificates, plain, secure)
  end
end
